### PR TITLE
promql: add sort_by_label and sort_by_label_desc functions

### PR DESCRIPTION
**Pull request by galexrt**
This adds functions to sort a vector by its label value.

Based on https://github.com/prometheus/prometheus/pull/1533

Signed-off-by: Alexander Trost <galexrt@googlemail.com>

***

There was some discussion around adding this functionality to PromQL/ Prometheus in the original PR #1533, though based on the various comments from users and now myself stumbling upon this while trying to visualize certain metrics sorted by label value, I would gladly see this topic revisited.

### CODE REVIEW

**Comment by roidelapluie** [comment]
How difficult would it be to make this a vararg function to be able to sort by multiple labels?

`sort_by_label(up, "job","instance")`

**Comment by galexrt** [comment]
@roidelapluie I can look into making it into a vararg function to be able to sort by multiple labels. Is there an existing function that already accepts a list of strings as an argument that I can look at?

**Comment by roidelapluie** [comment]
`label_join()`

**Comment by galexrt** [comment]
@roidelapluie The `sort_by_label` and `sort_by_label_desc` functions now accept a list of labels and is sorting the vector accordingly (see `functions.tests`).

File: docs/querying/functions.md [hunk 1]
```diff
@@ -401,6 +401,14 @@ in ascending order.
 
 Same as `sort`, but sorts in descending order.
 
+## `sort_by_label()`
+
+`sort_by_label(v instant-vector, label string, ...)` returns vector elements sorted by their label values and sample value in case of label values being equal, in ascending order.
```
Comment (juliusv) [inline docs/querying/functions.md:406]: Hmm, I wonder if we shouldn't keep this completely independent from the sample value and instead use other label values to break the sorting tie? That way, sort results would be stable over time, given the same set of time series.

File: promql/functions.go [hunk 2]
```diff
@@ -284,6 +284,24 @@ func funcSortDesc(vals []parser.Value, args parser.Expressions, enh *EvalNodeHel
 	return Vector(byValueSorter)
 }
 
+// === sort_by_label(vector parser.ValueTypeVector, label parser.ValueTypeString...) Vector ===
+func funcSortByLabel(vals []parser.Value, args parser.Expressions, enh *EvalNodeHelper) Vector {
+	// Incase the labels are the same, NaN should sort to the bottom, so take
```
Comment (juliusv) [inline promql/functions.go:289]: "Incase" -> "In case"

File: promql/functions.go [hunk 3]
```diff
@@ -1218,6 +1238,78 @@ func (s *vectorByReverseValueHeap) Pop() interface{} {
...
+	labels []string
+}
+
+func (s vectorByLabelHeap) Len() int {
+	return len(s.vector)
+}
+
+func (s vectorByLabelHeap) Less(i, j int) bool {
+	// Iterate over each given label
+	for k := 0; k < len(s.labels); k++ {
```
Comment (juliusv) [inline promql/functions.go:1252]: This would be more Go-y:

```suggestion
	for _, label := range s.labels {
```

File: docs/querying/functions.md (hunk 1 above)
Comment (roidelapluie) [reply docs/querying/functions.md:406]: I agree with Julius. We could also decide not to sort at all in that case, so you can do sort_by_label(sort(up),"job")

Comment (juliusv) [reply docs/querying/functions.md:406]: Yeah, that sounds even better.

Comment (galexrt) [reply docs/querying/functions.md:406]: So I should remove the "fallback" to comparing the value in the `sort_by_label` functions? This might make the test unstable, though I have to re-check if that happens without this "fallback" compare.

Comment (roidelapluie) [reply docs/querying/functions.md:406]: The tests should be stable even without this

File: promql/functions.go [hunk 4]
```diff
@@ -1218,6 +1238,78 @@ func (s *vectorByReverseValueHeap) Pop() interface{} {
...
+		lv2 := s.vector[j].Metric.Get(s.labels[k])
+		// 0 if a == b, -1 if a < b, and +1 if a > b.
+		switch strings.Compare(string(lv1), string(lv2)) {
+		case -1:
+			return true
+		case 0:
+			continue
+		case 1:
+			return false
+		}
```
Comment (roidelapluie) [inline promql/functions.go:1263]: we need a return at the end

File: docs/querying/functions.md (hunk 1 above)
Comment (roidelapluie) [reply docs/querying/functions.md:406]: we should make it clear that labels are sorted lexicographically

**Comment by bboreham** [comment]
Note that range queries are always sorted by label sets, right at the end:
https://github.com/prometheus/prometheus/blob/8b863c42dd956d35d18a7a0b39c89c86adf7cebf/promql/engine.go#L722

Perhaps you could mention in the documentation this is intended for instant queries.

**Comment by juliusv** [comment]
> Note that range queries are always sorted by label sets, right at the end:
https://github.com/prometheus/prometheus/blob/8b863c42dd956d35d18a7a0b39c89c86adf7cebf/promql/engine.go#L722
>
> Perhaps you could mention in the documentation this is intended for instant queries.

That same comment would apply to `sort()` and `sort_desc()` as well. In general, sorting is a no-op for range queries, as output points are always sorted by the X and Y axis there (unless someone cares about a specific legend item sorting, which I haven't heard yet).

I think we can either omit that comment on the new functions as well, or add a comment to the existing sort functions too. Something like:

> Please note that sorting functions only affect the results of instant queries, as range query results always have a fixed output ordering.

**Comment by bboreham** [comment]
I think it would be easy to work out that `sort` (by value) is not a good idea over a range, where different series can be the highest value at different points in time. 

Whereas `sort_by_label` is something you might reasonably want for a range query, e.g. to change the order in which lines are arranged in a dashboard. Hence it should be documented that it won't do what you asked.

**Comment by juliusv** [comment]
Ok yeah, that makes sense. Or we could actually make it work for range queries as well, but I doubt that's worth it. So then let's just add a comment for the new functions.

File: docs/querying/functions.md (hunk 1 above)
Comment (galexrt) [reply docs/querying/functions.md:406]: @roidelapluie Without the value comparison the test fails no matter in what order the result in the test(s) are placed.

Comment (roidelapluie) [reply docs/querying/functions.md:406]: That is strange. The order should be predictible.

Comment (galexrt) [reply docs/querying/functions.md:406]: It seems that it is not predictable as every X run the test fails and complains about the order of the result.

How should we approach this that we can hopefully get this functionality integrated?

File: promql/functions.go [hunk 5]
```diff
@@ -1218,6 +1238,78 @@ func (s *vectorByReverseValueHeap) Pop() interface{} {
 	return el
 }
 
+type vectorByLabelHeap struct {
```
Comment (bboreham) [inline promql/functions.go:1241]: What does the word "Heap" denote here?

File: promql/functions.go [hunk 6]
```diff
@@ -284,6 +284,24 @@ func funcSortDesc(vals []parser.Value, args parser.Expressions, enh *EvalNodeHel
...
+	sort.Sort(sort.Reverse(byLabelSorter))
+	return byLabelSorter.vector
+}
+
+// === sort_by_label_desc(vector parser.ValueTypeVector, label parser.ValueTypeString...) Vector ===
+func funcSortByLabelDesc(vals []parser.Value, args parser.Expressions, enh *EvalNodeHelper) Vector {
+	// Incase the labels are the same, NaN should sort to the bottom, so take
+	// ascending sort with NaN first and reverse it.
+	byLabelSorter := vectorByLabelHeap{vector: vals[0].(Vector), labels: stringSliceFromArgs(args[1:])}
+	sort.Sort(sort.Reverse(byLabelSorter))
```
Comment (bboreham) [inline promql/functions.go:301]: `slices.SortFunc` is probably neater now.

**Review by bboreham — COMMENTED** [review]
Hi, we discussed this PR at the [Prometheus dev summit](https://docs.google.com/document/d/11LC3wJcVk00l8w5P3oLQ-m3Y37iom6INAMEu2ZAGIIE/edit?usp=sharing) today; we would like to move forward, but would like to ask you to put these functions behind a feature flag.

I had a couple of comments about the implementation too.

**Comment by roidelapluie** [comment]
cc @galexrt are you still willing to work on this, rebase this pull request and put the functions behind a feature flag? Thanks!

**Comment by galexrt** [comment]
> cc @galexrt are you still willing to work on this, rebase this pull request and put the functions behind a feature flag? Thanks!

Oh that totally fell of my radar, yes I am willing to work on that. Do you have a link to a feature flag example somewhere that I can use as an example?

**Comment by roidelapluie** [comment]
You can look at the engine ops EnableNegativeOffset which used to be a feature flag

**Comment by galexrt** [comment]
> You can look at the engine ops EnableNegativeOffset which used to be a feature flag

So should I dynamically add the `sort_by_label` and `sort_by_label_desc` funcs to the `FunctionsList` map on start when the engine is created in `promql/engine.go`?

File: docs/querying/functions.md (hunk 1 above)
Comment (galexrt) [reply docs/querying/functions.md:558]: From my perspective, the fix here is to sort the values using the `sort`/`sort_desc` functions and then do the sort by label(s).

**Comment by roidelapluie** [comment]
I think a proper error saying that the feature needs to be enabled if the feature flag is not set would be better. However, I don't know if we can properly error from promql functions so we might need to be creative.

**Comment by galexrt** [comment]
@roidelapluie I think I have found a way to check the feature flag in the evaluator. Please take another look at the latest code changes, thanks!

**Comment by galexrt** [comment]
@roidelapluie I have updated the PR to resolve the conflicts. Any news on this?

File: docs/feature_flags.md [hunk 7]
```diff
@@ -127,10 +127,18 @@ corresponding native histogram. However, if a native histogram is present,
 Prometheus will ignore the corresponding conventional histogram, with the
 notable exception of exemplars, which are always ingested.
 
+
```
Comment (juliusv) [inline docs/feature_flags.md:130]: Unrelated new empty line?

File: promql/engine.go [hunk 8]
```diff
@@ -1386,6 +1394,8 @@ func (ev *evaluator) eval(expr parser.Expr) (parser.Value, annotations.Annotatio
 			if ok {
 				return ev.rangeEvalTimestampFunctionOverVectorSelector(vs, call, e)
 			}
+		} else if (e.Func.Name == "sort_by_label" || e.Func.Name == "sort_by_label_desc") && !ev.enableSortByLabel {
+			ev.error(fmt.Errorf("sort_by_label and sort_by_label_desc feature flag not enabled"))
```
Comment (juliusv) [inline promql/engine.go:1398]: This makes it sound as if `sort_by_label` and `sort_by_label_desc` are the feature flag names, how about something like this:

```promql
sort_by_label() and sort_by_label_desc() require the "promql-sort-by-label" feature flag to be set
```

**Review by juliusv — COMMENTED** [review]
Thank you, looks pretty good to me now!

It seems like the main outstanding question is around sort stability and whether we should resort to:

* Sorting by sample value
* Sorting by the remaining label set
* Do nothing

I would be for sorting by the remaining label set if that is easily doable (hopefully just a small extension of the SortFunc for the `0` string comparison case? Sorting by sample value seems too unrelated to the label sorting and is also not necessarily always stable, as two series can and often do have the same sample value.

File: documentation/examples/prometheus.yml [hunk 9]
```diff
@@ -20,10 +20,10 @@ rule_files:
...
   # The job name is added as a label `job=<job_name>` to any timeseries scraped from this config.
-  - job_name: "prometheus"
+  - job_name: "extended-ceph-exporter"
 
     # metrics_path defaults to '/metrics'
     # scheme defaults to 'http'.
 
     static_configs:
-      - targets: ["localhost:9090"]
+      - targets: ["localhost:9138"]
```
Comment (juliusv) [inline documentation/examples/prometheus.yml:29]: Looks like some other unrelated changes crept into the PR here now?

Comment (galexrt) [reply documentation/examples/prometheus.yml]: I have removed this unrelated change, PTAL

**Comment by juliusv** [comment]
Thanks! I see the tests failing with this error currently:

```
--- FAIL: TestEvaluations (5.87s)
    --- FAIL: TestEvaluations/testdata/functions.test (1.20s)
        test.go:105: 
            	Error Trace:	/__w/prometheus/prometheus/promql/test.go:105
            	            				/__w/prometheus/prometheus/promql/test.go:83
            	Error:      	Received unexpected error:
            	            	error in eval sort_by_label(http_requests, "instance", "group") (line 498): expected metric {__name__="http_requests", group="production", instance="0", job="api-server"} with [100.000000] at position 3 but was at 1
            	Test:       	TestEvaluations/testdata/functions.test
```

File: documentation/examples/prometheus.yml [hunk 10]
```diff
@@ -20,7 +20,7 @@ rule_files:
 # Here it's Prometheus itself.
 scrape_configs:
   # The job name is added as a label `job=<job_name>` to any timeseries scraped from this config.
-  - job_name: "prometheus"
+  - job_name: "extended-ceph-exporter"
```
Comment (juliusv) [inline documentation/examples/prometheus.yml:23]: I still see this unrelated change in here?

Comment (galexrt) [reply documentation/examples/prometheus.yml]: Fixed, all unrelated changes should be gone now

**Comment by galexrt** [comment]
> Thanks! I see the tests failing with this error currently:
> 
> ```
> --- FAIL: TestEvaluations (5.87s)
>     --- FAIL: TestEvaluations/testdata/functions.test (1.20s)
>         test.go:105: 
>             	Error Trace:	/__w/prometheus/prometheus/promql/test.go:105
>             	            				/__w/prometheus/prometheus/promql/test.go:83
>             	Error:      	Received unexpected error:
>             	            	error in eval sort_by_label(http_requests, "instance", "group") (line 498): expected metric {__name__="http_requests", group="production", instance="0", job="api-server"} with [100.000000] at position 3 but was at 1
>             	Test:       	TestEvaluations/testdata/functions.test
> ```

I have changed the sorting logic slightly, and it should now work better when more than 2 labels are involved.
To note sorting with 3 or more labels is "weird" because if the first label doesn't it doesn't compare the next label(s), added some test cases for that as well.

PTAL

**Review by roidelapluie — APPROVED** [review]

**Comment by roidelapluie** [comment]
Thanks!

**Comment by pstibrany** [comment]
It's amazing to see this merged, 9 years after first https://github.com/prometheus/prometheus/pull/1533 :) Thanks @galexrt and all the reviewers! 🎉

**Comment by bobrik** [comment]
Straightforward string sorting yield very questionable results for humans:

image

---

I urge you to consider using natural sorting instead.

Computers can sort by label however they want even without using `sort_by_label`, but humans need a little help.

See these PRs that added natural sorting to Grafana:

* https://github.com/grafana/grafana/pull/78024
* https://github.com/grafana/grafana/pull/78494

**Comment by juliusv** [comment]
@bobrik Agreed, good point! It's not yet too late to change things, since the new sorting functions are added under an experimental feature flag. Do you want to send a PR?

**Comment by bobrik** [comment]
Here's my attempt: https://github.com/prometheus/prometheus/pull/13411.
